conclave-enclave/src/main/kotlin/com/r3/conclave/enclave/internal/GramineEnclaveEnvironment.kt

```diff
@@ -28,103 +31,146 @@ class GramineEnclaveEnvironment(
+            //    array that it is filled by Gramine (which communicates with the quoting enclave in background).
+            val quotingEnclaveInfoBytes = quotingEnclaveInfo?.bytes ?: Cursor.allocate(SgxTargetInfo).bytes
+            createReport(ByteCursor.wrap(SgxTargetInfo, quotingEnclaveInfoBytes), reportData)
+            val signedQuoteBytes = readSignedQuote()
+            Cursor.slice(SgxSignedQuote, ByteBuffer.wrap(signedQuoteBytes))
+        }
     }
 
-    /** Generate simulated mrenclave value by hashing the enclave fat-jar */
-    private val mrenclave: ByteArray by lazy {
-        val digest = MessageDigest.getInstance("SHA-256")
+    override fun sealData(toBeSealed: PlaintextAndEnvelope): ByteArray {
+        //  TODO: Gramine filesystem support
+        return EnclaveUtils.sealData(aesSealingKey, toBeSealed)
+    }
 
-        val buffer = ByteArray(65536)
-        var bytesRead: Int
+    override fun unsealData(sealedBlob: ByteBuffer): PlaintextAndEnvelope {
+        //  TODO: Gramine filesystem support
+        return EnclaveUtils.unsealData(aesSealingKey, sealedBlob)
+    }
 
-        enclaveClass.protectionDomain.codeSource.location.openStream().use {
-            bytesRead = it.read(buffer)
-            while (bytesRead >= 0) {
-                digest.update(buffer, 0, bytesRead)
-                bytesRead = it.read(buffer)
-            }
+    override fun getSecretKey(keyRequest: ByteCursor<SgxKeyRequest>): ByteArray {
+        //  TODO: Gramine filesystem support
+        //  Note that we will probably not need to pass around the local secret key
+        //    as this one will be embedded in the manifest
+        //  https://gramine.readthedocs.io/en/stable/manifest-syntax.html#encrypted-files
+        return byteArrayOf()
+    }
+
+    override fun setupFileSystems(
+        inMemoryFsSize: Long,
+        persistentFsSize: Long,
+        inMemoryMountPath: String,
+        persistentMountPath: String,
+        encryptionKey: ByteArray
+    ) {
+        //  TODO: Gramine filesystem support
+    }
+
+    private fun retrieveReport(targetInfoBytes: ByteArray, userReportDataBytes: ByteArray): ByteArray {
+        writeTargetInfo(targetInfoBytes)
+        writeUserReportData(userReportDataBytes)
+        return readReport()
+    }
+
+    private fun readReport(): ByteArray {
+        return FileInputStream("/dev/attestation/report").use {
+            it.readBytes()
         }
+    }
 
-        digest.digest()
+    private fun readSignedQuote(): ByteArray {
+        return FileInputStream("/dev/attestation/quote").use {
+            it.readBytes()
+        }
     }
 
-    private val aesSealingKey by lazy(LazyThreadSafetyMode.NONE) {
-        digest("SHA-256") { update(enclaveClass.name.toByteArray()) }.copyOf(16)
+    private fun writeUserReportData(data: ByteArray) {
+
+        return FileOutputStream("/dev/attestation/user_report_data").use {
+            it.write(data)
+        }
```

Comment:
`Paths.get("/dev/attestation/user_report_data").writeBytes(data)`

Reply:
Always forget, doing it